File: test_app.py
#!/usr/bin/env python3
"""
Weather Application Playwright Test

Tests:
1. Page loads successfully
2. API calls work (OpenWeatherMap integration)
3. UI elements are present (search, current weather, forecast, favorites)
4. Search functionality
5. Favorites functionality
6. Mobile responsiveness (iPhone 12 dimensions)
"""

from playwright.sync_api import sync_playwright
import sys


def test_weather_app():
    """Run comprehensive tests on the weather application."""
    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        context = browser.new_context(viewport={'width': 1280, 'height': 720})
        page = context.new_page()

        errors = []
        console_errors = []

        # Capture console errors
        page.on('console', lambda msg: capture_console(msg, console_errors))
        page.on('pageerror', lambda exc: errors.append(str(exc)))

        print("=" * 60)
        print("Weather Application Test")
        print("=" * 60)

        # Test 1: Page loads
        print("\n[TEST 1] Loading page...")
        try:
            page.goto('http://localhost:5173', timeout=30000)
            page.wait_for_load_state('networkidle', timeout=30000)
            print("  [OK] Page loaded successfully")

            # Wait a bit for React to hydrate
            page.wait_for_timeout(2000)
        except Exception as e:
            print(f"  [FAIL] Failed to load page: {e}")
            return False

        # Test 2: Check for main UI elements
        print("\n[TEST 2] Checking UI elements...")
        ui_tests = [
            ('header, nav, [class*="header"]', 'Header'),
            ('input[placeholder*="城市" i], input[placeholder*="搜索" i]', 'Search input'),
            ('text=天气, text=Weather', 'Weather section'),
            ('text=预报, text=forecast', 'Forecast section'),
            ('svg, [class*="icon"]', 'Icons'),
        ]

        for selector, name in ui_tests:
            try:
                if 'text=' in selector:
                    # Handle text selector
                    locator = page.get_by_text(selector.replace('text=', '').strip('[]'))
                    # Wait for it to be available
                    locator.first.wait_for(timeout=5000)
                else:
                    locator = page.locator(selector.split(',')[0].strip())

                if locator.count() > 0:
                    print(f"  [OK] {name} found")
                else:
                    print(f"  [WARN] {name} not found (may be expected)")
            except Exception as e:
                print(f"  [WARN] {name}: {e}")

        # Test 3: Check console for errors
        print("\n[TEST 3] Console errors check...")
        if console_errors:
            print(f"  Found {len(console_errors)} console errors:")
            for err in console_errors[:5]:  # Show first 5
                print(f"    - {err}")
        else:
            print("  [OK] No console errors")

        # Test 4: Page errors
        print("\n[TEST 4] Page errors check...")
        if errors:
            print(f"  [FAIL] Found {len(errors)} page errors:")
            for err in errors[:5]:
                print(f"    - {err}")
        else:
            print("  [OK] No page errors")

        # Test 5: Take screenshot for visual verification
        print("\n[TEST 5] Taking screenshot...")
        try:
            page.screenshot(path='/tmp/weather_app_screenshot.png', full_page=True)
            print("  [OK] Screenshot saved to /tmp/weather_app_screenshot.png")
        except Exception as e:
            print(f"  [FAIL] Screenshot failed: {e}")

        # Test 6: Check page content
        print("\n[TEST 6] Page content check...")
        content = page.content()
        body_text = page.locator('body').inner_text()

        # Check for common weather-related content
        checks = [
            ('body_text', '天气' in body_text or 'Weather' in body_text or '°' in body_text),
            ('body_text', '北京' in body_text or '上海' in body_text or body_text.strip() != ''),
        ]

        for name, passed in checks:
            if passed:
                print(f"  [OK] {name}: contains expected content")
            else:
                print(f"  [WARN] {name}: may not contain expected content")

        # Test 7: Mobile responsiveness (iPhone 12 dimensions)
        print("\n[TEST 7] Mobile responsiveness test (iPhone 12 - 390x844)...")
        try:
            # Create mobile context
            mobile_context = browser.new_context(
                viewport={'width': 390, 'height': 844},
                is_mobile=True,
                has_touch=True
            )
            mobile_page = mobile_context.new_page()

            # Capture mobile console errors
            mobile_errors = []
            mobile_console_errors = []
            mobile_page.on('console', lambda msg: capture_console(msg, mobile_console_errors))
            mobile_page.on('pageerror', lambda exc: mobile_errors.append(str(exc)))

            # Navigate to page
            mobile_page.goto('http://localhost:5173', timeout=30000)
            mobile_page.wait_for_load_state('networkidle', timeout=30000)
            mobile_page.wait_for_timeout(2000)

            # Check if sidebar overlay/mask exists and works properly
            # The overlay should be z-40 and sidebar z-50
            overlay = mobile_page.locator('[class*="fixed"][class*="inset-0"][class*="bg-black/"]').first
            sidebar = mobile_page.locator('aside').first

            # Check for overlay (should exist when sidebar is visible)
            overlay_count = mobile_page.locator('[class*="fixed"][class*="inset-0"][class*="z-40"]').count()
            sidebar_z50 = mobile_page.locator('[class*="z-50"]').count()

            print(f"  [INFO] Overlay elements found: {overlay_count}")
            print(f"  [INFO] z-50 elements found: {sidebar_z50}")

            # Take mobile screenshot
            mobile_page.screenshot(path='/tmp/weather_app_mobile.png', full_page=True)
            print("  [OK] Mobile screenshot saved to /tmp/weather_app_mobile.png")

            # Check for console errors on mobile
            if mobile_console_errors:
                print(f"  [WARN] Mobile console errors: {len(mobile_console_errors)}")
            else:
                print("  [OK] No mobile console errors")

            if mobile_errors:
                print(f"  [FAIL] Mobile page errors: {len(mobile_errors)}")
                for err in mobile_errors[:3]:
                    print(f"    - {err}")
            else:
                print("  [OK] No mobile page errors")

            mobile_context.close()
            print("  [OK] Mobile test completed")

        except Exception as e:
            print(f"  [FAIL] Mobile test failed: {e}")

        # Test 8: Network requests (API calls)
        print("\n[TEST 8] Network request monitoring...")

        # Start waiting for responses
        response_promises = []

        # Navigate again to capture requests
        page.goto('http://localhost:5173')
        page.wait_for_timeout(3000)

        # Check if any API calls were made (console log check)
        if console_errors:
            api_errors = [e for e in console_errors if 'fetch' in e.lower() or 'api' in e.lower()]
            if api_errors:
                print(f"  [WARN] API-related errors found: {len(api_errors)}")
            else:
                print("  [OK] No API-related errors in console")

        # Summary
        print("\n" + "=" * 60)
        print("Test Summary")
        print("=" * 60)
        print(f"Console errors: {len(console_errors)}")
        print(f"Page errors: {len(errors)}")

        if errors:
            print("\n[FAIL] TESTS FAILED - Page errors detected")
            browser.close()
            return False

        if len([e for e in console_errors if 'error' in e.lower() and 'favicon' not in e.lower()]) > 3:
            print("\n[FAIL] TESTS FAILED - Too many console errors")
            browser.close()
            return False

        print("\n[SUCCESS] TESTS PASSED - Application loaded successfully")
        browser.close()
        return True


def capture_console(msg, errors_list):
    """Capture console messages, focusing on errors."""
    if msg.type == 'error':
        # Ignore favicon errors
        if 'favicon' not in msg.text.lower():
            errors_list.append(msg.text)


if __name__ == '__main__':
    success = test_weather_app()
    sys.exit(0 if success else 1)

File: test-app.mjs
import { chromium } from 'playwright';

async function testWeatherApp() {
  console.log('=== 天气应用测试 ===\n');
  
  const browser = await chromium.launch({ headless: true });
  const page = await browser.newPage();
  
  const errors = [];
  page.on('console', msg => {
    if (msg.type() === 'error') {
      errors.push(msg.text());
    }
  });

  try {
    // 1. 打开应用
    console.log('1. 打开应用...');
    await page.goto('http://localhost:5175', { waitUntil: 'networkidle', timeout: 30000 });
    await page.waitForTimeout(3000);
    console.log('   ✅ 页面加载成功\n');

    // 2. 检查基本元素
    console.log('2. 检查页面元素...');
    
    // 检查 Header
    const header = await page.$('header');
    console.log(`   ${header ? '✅' : '❌'} Header: ${header ? '存在' : '缺失'}`);
    
    // 检查侧边栏（不应该有"天气"标题）
    const sidebarTitle = await page.$('text=天气');
    const sidebarTitleCount = sidebarTitle ? await sidebarTitle.count() : 0;
    console.log(`   ${sidebarTitleCount <= 1 ? '✅' : '⚠️'} 侧边栏标题: ${sidebarTitleCount <= 1 ? '已移除或仅在Header' : '还有残留'}`);
    
    // 检查搜索按钮
    const searchBtn = await page.getByRole('button', { name: /搜索城市/ });
    console.log(`   ${await searchBtn.isVisible().catch(() => false) ? '✅' : '❌'} 搜索按钮: 可见`);
    
    // 检查刷新按钮
    const refreshBtn = await page.getByRole('button', { name: /刷新/ });
    console.log(`   ${await refreshBtn.isVisible().catch(() => false) ? '✅' : '❌'} 刷新按钮: 可见`);
    
    // 检查主题切换
    const themeBtn = await page.locator('header button').last();
    console.log(`   ${await themeBtn.isVisible().catch(() => false) ? '✅' : '❌'} 主题切换: 可见\n`);

    // 3. 测试搜索功能
    console.log('3. 测试搜索功能...');
    const searchButton = await page.locator('button:has-text("搜索城市")');
    if (await searchButton.isVisible().catch(() => false)) {
      await searchButton.click();
      await page.waitForTimeout(500);
      
      // 检查搜索弹窗
      const searchInput = await page.locator('input[placeholder*="搜索"]');
      if (await searchInput.isVisible().catch(() => false)) {
        console.log('   ✅ 搜索弹窗打开');
        
        // 测试输入
        await searchInput.fill('上海');
        await page.waitForTimeout(1000);
        console.log('   ✅ 搜索框可以输入文字');
        
        // 关闭搜索弹窗
        const cancelBtn = await page.getByRole('button', { name: '取消' });
        if (await cancelBtn.isVisible().catch(() => false)) {
          await cancelBtn.click();
          console.log('   ✅ 关闭搜索弹窗');
        }
      }
    }
    console.log('');

    // 4. 检查预报卡片
    console.log('4. 检查预报卡片...');
    const forecastCards = await page.locator('.neumorph-card').count();
    console.log(`   ✅ 找到 ${forecastCards} 个卡片\n`);

    // 5. 检查温度趋势图表
    console.log('5. 检查温度趋势图表...');
    const chart = await page.locator('.recharts-wrapper, svg').first();
    console.log(`   ${await chart.isVisible().catch(() => false) ? '✅' : '❌'} 温度趋势图表: 可见\n`);

    // 6. 检查控制台错误
    console.log('6. 控制台错误检查...');
    if (errors.length > 0) {
      console.log(`   ⚠️ 发现 ${errors.length} 个错误:`);
      errors.slice(0, 3).forEach((e, i) => console.log(`   ${i+1}. ${e.substring(0, 100)}`));
    } else {
      console.log('   ✅ 无控制台错误\n');
    }

    console.log('=== 测试完成 ===');

  } catch (error) {
    console.error('❌ 测试失败:', error.message);
  } finally {
    await browser.close();
  }
}

testWeatherApp();
